offer/retrieval union content types splitted into separate types (#262)
ethereum/portal-network-specs: state-network.md

selector               := 0x20

content                := Container(proof: StateWitness, block_hash: Bytes32)
content_key            := selector + SSZ.serialize(account_trie_node_key)
content_id             := sha256(content_key)
```

#### Account Trie Node for Retrival

This data type represents a node from the main state trie when retrieved from another node by explicit request.

```
account_trie_node_key  := Container(path: Nibbles, node_hash: Bytes32)
selector               := 0x21

content                := Container(node: WitnessNode)
content_key            := selector + SSZ.serialize(account_trie_node_key)
content_id             := sha256(content_key)
```

#### Contract Trie Node for Offer

This data type represents a node from an individual contract storage trie when offered via gossip.

```
storage_trie_node_key  := Container(address: Address, path: Nibbles, node_hash: Bytes32)
selector               := 0x22

content                :=  Container(proof: StorageWitness, block_hash: Bytes32)
content_key            :=  selector + SSZ.serialize(storage_trie_node_key)
content_id             :=  sha256(content_key)
```

#### Contract Trie Node for Retrival

> This data type represents a node from an individual contract storage trie when retrieved from another node by explicit request.

Reviewer: ```suggestion
This data type represents a node from an individual contract storage trie when retrieved from another node via FINDCONTENT/FOUNDCONTENT.
```

Maybe these are better with explicit names for the message types?

Not going to make this suggestion for all of these (if you agree with this change it should be replicated to each of these definitions)

```
storage_trie_node_key  := Container(address: Address, path: Nibbles, node_hash: Bytes32)
selector               := 0x23

content                :=  Container(node: WitnessNode)
content_key            :=  selector + SSZ.serialize(storage_trie_node_key)
content_id             :=  sha256(content_key)
```

> #### Contract Code for Offer

Reviewer: ```suggestion
#### Contract Code: OFFER/ACCEPT
```

What do you think about this.  It's a minor formatting change but maybe is "cleaner" when these things are rendered in a table-of-contents style display.

Problematic!

```
contract_code_key      := Container(address: Address, code_hash: Bytes32)
selector               := 0x24

content_for_offer      := Container(code: ByteList, account_proof: StateWitness, block_hash: Bytes32)
content_for_retrieval  := Container(code: ByteList)
content_key            := selector + SSZ.serialize(contract_code_key)
content_id             := sha256(content_key)
```

#### Contract Code for Retrival

This data type represents the bytecode for a contract when retrieved from another node by explicit request.

> NOTE: Because CREATE2 opcode allows for redeployment of new code at an existing address, we MUST randomly distribute contract code storage across the DHT keyspace to avoid hotspots developing in the network for any contract that has had many different code deployments.  Were we to use the path based *high-bits* approach for computing the content-id, it would be possible for a single location in the network to accumulate a large number of contract code objects that all live in roughly the same space.
Problematic!

```
contract_code_key      := Container(address: Address, code_hash: Bytes32)

> selector               := 0x25

Reviewer: I don't think we can have the selector be different for these types (it seems that you've made it so that the selector is different for the offer/accept vs find/found content types.  That means that they actually end up with different content-id values.  We need them to have the same content-id which means they have to have the same selector across the offer/accept and find/found types.

#### Account Trie Node

This data types represent a node from the main state trie.

```
account_trie_node_key  := Container(path: Nibbles, node_hash: Bytes32)
selector               := 0x20

content_key            := selector + SSZ.serialize(account_trie_node_key)
content_id             := sha256(content_key)
```

##### Account Trie Node: OFFER/ACCEPT

> This type is used when content offered via OFFER/ACCEPT.

Reviewer 2: I would be more explicit here and use `MUST`.

```
content_for_offer      := Container(proof: StateWitness, block_hash: Bytes32)
```

##### Account Trie Node: FINDCONTENT/FOUNDCONTENT

> This type is used when content retrieved from another node via FINDCONTENT/FOUNDCONTENT.

Reviewer 2: Would use `SHOULD` here. (Or perhaps also `MUST`).

```
content_for_retrieval := Container(node: WitnessNode)
```

#### Contract Trie Node

This data types represent a node from an individual contract storage trie.

```
storage_trie_node_key  := Container(address: Address, path: Nibbles, node_hash: Bytes32)
selector               := 0x21

content_key            := selector + SSZ.serialize(storage_trie_node_key)
content_id             := sha256(content_key)
```

##### Contract Trie Node: OFFER/ACCEPT

> This type is used when content offered via OFFER/ACCEPT.

Reviewer 2: idem MUST

```
content_for_offer      := Container(proof: StorageWitness, block_hash: Bytes32)
```

##### Contract Trie Node: FINDCONTENT/FOUNDCONTENT

> This type is used when content retrieved from another node via FINDCONTENT/FOUNDCONTENT.

Reviewer 2: idem SHOULD/MUST

```
content_for_retrieval  := Container(node: WitnessNode)
```

#### Contract Code

> This data types represent the bytecode for a contract.

Reviewer 2: These

contract_code_key      := Container(address: Address, code_hash: Bytes32)
selector               := 0x22

content_key            := selector + SSZ.serialize(contract_code_key)
content_id             := sha256(content_key)
```

##### Contract Code: OFFER/ACCEPT

This types is used when content offered via OFFER/ACCEPT.

```
content_for_offer      := Container(code: ByteList, account_proof: StateWitness, block_hash: Bytes32)
```

##### Contract Code: FINDCONTENT/FOUNDCONTENT

> This type is used when content retrieved from another node via FINDCONTENT/FOUNDCONTENT.

Reviewer 2: Idem above
